### PR TITLE
Send message

### PR DESCRIPTION
**Pull request by Laursen79**
This pull request includes significant changes to the `client` module, primarily focusing on refactoring the `Client` struct, enhancing encryption and decryption processes, and improving error handling. The most important changes are summarized below:

### Refactoring and Struct Changes:
* [`client/src/client.rs`](diffhunk://#diff-f74337de83c9232a0cc949a8d6d7390cf6e46b011b79383d9666ede90a902526L1-R51): Refactored the `Client` struct to use generic parameters for `StorageType` and `SignalServerAPI`, replacing the previous hardcoded types. This change aims to improve flexibility and reusability of the `Client` struct. [[1]](diffhunk://#diff-f74337de83c9232a0cc949a8d6d7390cf6e46b011b79383d9666ede90a902526L1-R51) [[2]](diffhunk://#diff-f74337de83c9232a0cc949a8d6d7390cf6e46b011b79383d9666ede90a902526L67-L84) [[3]](diffhunk://#diff-f74337de83c9232a0cc949a8d6d7390cf6e46b011b79383d9666ede90a902526L95-R108) [[4]](diffhunk://#diff-f74337de83c9232a0cc949a8d6d7390cf6e46b011b79383d9666ede90a902526L144-R134) [[5]](diffhunk://#diff-f74337de83c9232a0cc949a8d6d7390cf6e46b011b79383d9666ede90a902526L162-R174) [[6]](diffhunk://#diff-f74337de83c9232a0cc949a8d6d7390cf6e46b011b79383d9666ede90a902526L193-R357)

### Encryption and Decryption Enhancements:
* [`client/src/encryption.rs`](diffhunk://#diff-8966d253578f9a35205994056f1802222418f00d4f2f4dd29101c1e89081bea7L1-R44): Added new functions `pad_message` and `unpad_message` to handle message padding, and updated the `encrypt` and `decrypt` functions to improve the encryption process by including timestamp and error handling improvements. [[1]](diffhunk://#diff-8966d253578f9a35205994056f1802222418f00d4f2f4dd29101c1e89081bea7L1-R44) [[2]](diffhunk://#diff-8966d253578f9a35205994056f1802222418f00d4f2f4dd29101c1e89081bea7L43-R131) [[3]](diffhunk://#diff-8966d253578f9a35205994056f1802222418f00d4f2f4dd29101c1e89081bea7L185-R283)

### Error Handling Improvements:
* [`client/src/contact_manager.rs`](diffhunk://#diff-514d8684053ef86f2a59bcb2d718f408fba3f724480ed1640e2fd2642e654899L2-R5): Updated the `get_address` method in the `Contact` struct to return a custom `SignalClientError` instead of `SignalProtocolError` for better error specificity. [[1]](diffhunk://#diff-514d8684053ef86f2a59bcb2d718f408fba3f724480ed1640e2fd2642e654899L2-R5) [[2]](diffhunk://#diff-514d8684053ef86f2a59bcb2d718f408fba3f724480ed1640e2fd2642e654899L19-R26)

### Dependency Updates:
* [`client/Cargo.toml`](diffhunk://#diff-6a7ec327336edab1ae223d43992fdda7a8c6b5cad3cce55a8db0926c00448026R35-R37): Added new dependencies `derive_more` and `display` to enhance error handling and display functionalities.

These changes collectively aim to improve the flexibility, reliability, and maintainability of the `client` module.

### CODE REVIEW

File: '
Comment (Mikkel-Boje) [inline ':1]: Huh

File: client/src/client.rs [hunk 1]
```diff
@@ -193,11 +193,296 @@ impl<T: StorageType> Client<T> {
...
+        let to = match self.contact_manager.get_contact(service_id) {
+            Err(_) => {
+                self.contact_manager
+                    .add_contact(service_id, 1.into())
+                    .expect("Can add contact that does not exist yet");
+                let bundles = self
+                    .server_api
+                    .fetch_pre_key_bundles(service_id.service_id_string())
+                    .await
+                    .unwrap();
```
Comment (UnluckyBird) [inline client/src/client.rs:222]: expect?

File: client/src/client.rs [hunk 2]
```diff
@@ -193,11 +193,296 @@ impl<T: StorageType> Client<T> {
...
+
+                let mut device_ids = Vec::new();
+                for ref bundle in bundles {
+                    device_ids.push(bundle.device_id().unwrap());
```
Comment (UnluckyBird) [inline client/src/client.rs:226]: expect?

File: client/src/client.rs [hunk 3]
```diff
@@ -193,11 +193,296 @@ impl<T: StorageType> Client<T> {
...
+                            bundle.device_id().unwrap(),
+                        ),
+                        &mut self.storage.protocol_store.session_store,
+                        &mut self.storage.protocol_store.identity_key_store,
+                        bundle,
+                        SystemTime::now(),
+                        &mut OsRng,
+                    )
+                    .await
+                    .unwrap();
```
Comment (UnluckyBird) [inline client/src/client.rs:239]: expect?

File: client/src/client.rs [hunk 4]
```diff
@@ -193,11 +193,296 @@ impl<T: StorageType> Client<T> {
...
+
+        // Envelope contains a ciphertext message; a so called `encrypted [Content]`
+        let ciphertext_content = envelope.content();
+
+        // The content of the envelope is base64 encoded.
+        let bytes = BASE64_STANDARD
+            .decode(ciphertext_content)
+            .map_err(ReceiveMessageError::Base64DecodeError)?;
+
+        // The evelope contains information about which message type is received.
```
Comment (UnluckyBird) [inline client/src/client.rs:307]: envelope*

File: client/src/client.rs [hunk 5]
```diff
@@ -193,11 +193,296 @@ impl<T: StorageType> Client<T> {
...
+        let _type = match envelope.r#type() {
+            envelope::Type::Ciphertext => Ok(CiphertextMessageType::Whisper),
+            envelope::Type::PrekeyBundle => Ok(CiphertextMessageType::PreKey),
+            //7 => Ok(CiphertextMessageType::SenderKey),
```
Comment (UnluckyBird) [inline client/src/client.rs:311]: What does this mean?

File: client/src/encryption.rs [hunk 6]
```diff
@@ -1,37 +1,49 @@
...
+) -> Result<HashMap<DeviceId, (u32, CiphertextMessage)>, SignalClientError> {
+    let mut msgs = HashMap::new();
     for id in target.device_ids.clone() {
+        let reg_id = session_store
+            .load_session(&target.get_address(&id).unwrap())
+            .await
+            .unwrap()
+            .unwrap()
+            .remote_registration_id()
+            .unwrap();
```
Comment (UnluckyBird) [inline client/src/encryption.rs:29]: expect? 3x

File: client/src/client.rs [hunk 7]
```diff
@@ -193,11 +193,296 @@ impl<T: StorageType> Client<T> {
...
+                        CiphertextMessage::PlaintextContent(_) => {
+                            envelope::Type::PlaintextContent.into()
+                        }
+                    },
+                    destination_device_id: id.into(),
+                    destination_registration_id: msg.0,
+                    content: BASE64_STANDARD.encode(msg.1.serialize()),
+                })
+                .collect(),
+            online: false, // Should this be true?
```
Comment (UnluckyBird) [inline client/src/client.rs:280]: This should not be true

**Review by UnluckyBird — APPROVED** [review]
.expect("Probably works")

File: client/src/client.rs [hunk 8]
```diff
@@ -197,11 +204,296 @@ impl<T: StorageType> Client<T> {
...
+                    device_ids.push(bundle.device_id().unwrap());
+                    process_prekey_bundle(
+                        &ProtocolAddress::new(
+                            service_id.service_id_string(),
+                            bundle.device_id().unwrap(),
+                        ),
+                        &mut self.storage.protocol_store.session_store,
+                        &mut self.storage.protocol_store.identity_key_store,
+                        bundle,
+                        SystemTime::now(),
```
Comment (Mikkel-Boje) [inline client/src/client.rs:246]: Maybe save this in a variable outside the loop such that it shows the keys as being created at the same point in time.

File: client/src/client.rs [hunk 9]
```diff
@@ -197,11 +204,296 @@ impl<T: StorageType> Client<T> {
...
+        alice
+            .send_message("Hello, World!", &bob.aci.into())
+            .await
+            .unwrap();
+
+        let message = bob.receive_message().await.unwrap();
+
+        assert_eq!("Hello, World!".to_owned(), message);
+    }
+    */
```
Comment (Mikkel-Boje) [inline client/src/client.rs:479]: Remove

File: client/src/encryption.rs [hunk 10]
```diff
@@ -182,29 +240,47 @@ pub mod test {
...
         .await;
 
-        let msg_map = encrypt(&mut alice_store, bob, "Hello Bob".as_bytes())
-            .await
-            .unwrap();
+        let msg_map = encrypt(
+            &mut alice_store.identity_key_store,
+            &mut alice_store.session_store,
+            bob,
+            "Hello Bob".as_bytes(),
```
Comment (Mikkel-Boje) [inline client/src/encryption.rs:254]: Why a hardcoded str?

File: client/src/errors.rs [hunk 11]
```diff
@@ -2,43 +2,105 @@ use std::error::Error;
...
+            Self::ContactManagerError(err) => write!(f, "{err}"),
+            Self::RegistrationError(err) => write!(f, "{err}"),
+            Self::LoginError(err) => write!(f, "{err}"),
+            Self::SendMessageError(err) => write!(f, "{err}"),
+            Self::WebSocketError(err) => write!(f, "{err}"),
+            Self::DatabaseError(err) => write!(f, "{err}"),
+            Self::DotenvError(err) => write!(f, "{err}"),
+            Self::ReceiveMessageError(err) => write!(f, "{err}"),
+            Self::ServerRequestError(err) => write!(f, "{err}"),
+            Self::KeyError(err) => write!(f, "{err}"),
```
Comment (Mikkel-Boje) [inline client/src/errors.rs:66]: Self::ContactManagerError(err)
            | Self::RegistrationError(err)
            | Self::LoginError(err)
            | Self::SendMessageError(err)
            | Self::WebSocketError(err)
            | Self::DatabaseError(err)
            | Self::DotenvError(err)
            | Self::ReceiveMessageError(err)
            | Self::ServerRequestError(err)
            | Self::KeyError(err) => write!(f, "{err}"),

File: client/src/server.rs [hunk 12]
```diff
@@ -174,74 +233,91 @@ impl Server for ServerAPI {
...
+        if res.status().is_success() {
+            Ok(from_slice(
+                res.body_bytes()
+                    .await
+                    .map_err(|err| RegistrationError::BadResponse(format!("{err}")))
+                    .expect("There")
+                    .as_ref(),
+            )
+            .map_err(|err| RegistrationError::BadResponse(format!("{err}")))
+            .expect("Here"))
```
Comment (Mikkel-Boje) [inline client/src/server.rs:259]: Then just Unwrap, since unwrap also shows where

File: client/src/server.rs [hunk 13]
```diff
@@ -174,74 +233,91 @@ impl Server for ServerAPI {
...
-                .as_ref(),
-        )
-        .map_err(|_| RegistrationError::BadResponse)?)
+            .map_err(|_| RegistrationError::NoResponse)?;
+        if res.status().is_success() {
+            Ok(from_slice(
+                res.body_bytes()
+                    .await
+                    .map_err(|err| RegistrationError::BadResponse(format!("{err}")))
+                    .expect("There")
```
Comment (Mikkel-Boje) [inline client/src/server.rs:255]: Please

File: client/src/server.rs [hunk 14]
```diff
@@ -174,74 +233,91 @@ impl Server for ServerAPI {
...
+    async fn get_message(&mut self) -> Option<Envelope> {
+        let x = self.message_queue.recv().await?;
 
-        self.make_request(ReqType::Delete(payload), uri).await;
-        todo!("Handle the response")
-    }
+        let req = match x.request {
+            Some(x) => x,
+            None => return None,
+        };
```
Comment (Mikkel-Boje) [inline client/src/server.rs:297]:
```        
        let Some(req) = self.message_queue.recv().await?.request else {
            return None
        };
```

File: client/src/socket_manager.rs [hunk 15]
```diff
@@ -289,9 +289,9 @@ impl<T: WSStream<Message, tungstenite::Error> + std::fmt::Debug> SocketManager<T
                 }
                 web_socket_message::Type::Request => {
                     let _id = msg
-                        .response
+                        .request
                         .as_ref()
-                        .expect("No Response in response")
+                        .expect("No Request in response")
```
Comment (Mikkel-Boje) [inline client/src/socket_manager.rs:294]: Why caps on Request xd

File: client/src/server.rs [hunk 16]
```diff
@@ -322,8 +398,153 @@ impl Display for ServerRequestError {
...
+        }
+    }
+
+    impl MockBackend {
+        pub fn new(address: ProtocolAddress, state: Arc<Mutex<MockBackendState>>) -> Self {
+            Self { address, state }
+        }
+    }
+}
+*/
```
Comment (Mikkel-Boje) [inline client/src/server.rs:545]: Do we need this?

File: client/src/server.rs [hunk 17]
```diff
@@ -174,74 +233,92 @@ impl Server for ServerAPI {
...
-            "uuid": uuid
-        });
-        let uri = format!("{}/{}", DEVICE_URI, uuid);
+        match Envelope::decode(req.body()) {
+            Ok(e) => Some(e),
+            Err(e) => {
+                println!("Failed to decode envelope: {}", e);
+                None
+            }
+        }
```
Comment (Mikkel-Boje) [inline client/src/server.rs:305]:
```
        Envelope::decode(req.body()).ok().or_else(|err| {
            println!("Failed to decode envelope: {}", err);
        })
```

**Review by Mikkel-Boje — CHANGES_REQUESTED** [review]

File: client/src/client.rs (hunk 1 above)
Comment (Laursen79) [reply client/src/client.rs:233]: Det skal håndteres ordenligt på et tidspunkt

File: client/src/client.rs (hunk 2 above)
Comment (Laursen79) [reply client/src/client.rs:237]: Denne funktion retunerer Ok(device_id) altid, så jeg synes unwrap er fint

File: client/src/server.rs [hunk 18]
```diff
@@ -1,34 +1,154 @@
...
+
+    pub async fn get_message(&mut self) -> Option<Envelope> {
+        self.backend.get_message().await
+    }
+    pub fn create_auth_header(&mut self, aci: Aci, password: String, device_id: DeviceId) -> () {
+        self.backend.create_auth_header(aci, password, device_id);
+    }
+}
+
+pub struct SignalBackend {
```
Comment (MJHC) [inline client/src/server.rs:149]: backend backend

File: client/src/server.rs [hunk 19]
```diff
@@ -55,52 +175,14 @@ impl Display for ReqType {
...
-    async fn delete_client(&self, uuid: String) -> Result<(), Self::Error>;
-    async fn delete_device(&self, uuid: String) -> Result<(), Self::Error>;
-    async fn get_message(&mut self) -> Option<Envelope>;
-    fn create_auth_header(&mut self, aci: Aci, password: &str, device_id: DeviceId) -> ();
-    fn new() -> Self;
-}
-
-impl Server for ServerAPI {
-    type Error = SignalClientError;
+impl Backend for SignalBackend {
```
Comment (MJHC) [inline client/src/server.rs:178]: backend?

File: client/src/errors.rs (hunk 11 above)
Comment (Laursen79) [reply client/src/errors.rs:66]: Det skal gøres

File: client/src/server.rs [hunk 20]
```diff
@@ -1,34 +1,154 @@
...
+}
+
+#[allow(unused)]
+impl VerifiedSession {
+    pub fn session_id(&self) -> &String {
+        &self.session_id
+    }
+}
+
+pub trait Backend {
```
[the remark this answers is not in the document]
Comment (MJHC) [reply client/src/server.rs:43]: maybe this could be named ServerAPI

File: client/src/server.rs (hunk 13 above)
Comment (Laursen79) [reply client/src/server.rs:255]: Skal være ?

File: client/src/server.rs (hunk 14 above)
Comment (Laursen79) [reply client/src/server.rs:297]: ok

File: client/src/server.rs (hunk 17 above)
Comment (Laursen79) [reply client/src/server.rs:305]: ok

File: client/src/server.rs [hunk 21]
```diff
@@ -1,34 +1,154 @@
...
+        messages: SignalMessages,
+        service_id: &ServiceId,
+    ) -> Result<(), SignalClientError>;
+
+    async fn get_message(&mut self) -> Option<Envelope>;
+    fn create_auth_header(&mut self, aci: Aci, password: String, device_id: DeviceId) -> ();
+}
+
+#[derive(Debug)]
+pub struct ServerAPI<T: Backend> {
```
Comment (MJHC) [inline client/src/server.rs:87]: and this could be named Server

File: client/src/server.rs (hunk 16 above)
Comment (Laursen79) [reply client/src/server.rs:545]: Venter på alex Keymanager create_key_bundle

File: client/src/server.rs [hunk 22]
```diff
@@ -322,8 +398,153 @@ impl Display for ServerRequestError {
 }
 
 impl Error for ServerRequestError {}
+/*
+#[cfg(test)]
+pub mod server_api_test {
```
Comment (MJHC) [inline client/src/server.rs:403]: why is this commented?

File: client/src/socket_manager.rs (hunk 15 above)
Comment (Laursen79) [reply client/src/socket_manager.rs:294]: Lad os få en ordentlig fejlbesked

File: client/src/server.rs [hunk 23]
```diff
@@ -322,8 +398,153 @@ impl Display for ServerRequestError {
...
+    impl MockBackend {
+        pub fn new(address: ProtocolAddress, state: Arc<Mutex<MockBackendState>>) -> Self {
+            Self { address, state }
+        }
+    }
+}
+*/
 
-impl ServerAPI {
+impl SignalBackend {
```
Comment (MJHC) [inline client/src/server.rs:547]: there is many backend

File: client/src/socket_manager.rs [hunk 24]
```diff
@@ -25,7 +25,7 @@ use tokio_tungstenite::{
 };
 use tungstenite::protocol::frame::coding::CloseCode;
 
-fn rustls_cfg(ca_file_path: &str) -> Result<ClientConfig, String> {
+pub fn rustls_cfg(ca_file_path: &str) -> Result<ClientConfig, String> {
```
Comment (MJHC) [inline client/src/socket_manager.rs:28]: why is this pub?

**Review by MJHC — CHANGES_REQUESTED** [review]
please make an issue on the generic problem, and fix the confusing naming that is going on in `server.rs`

File: client/src/server.rs (hunk 13 above)
Comment (MJHC) [reply client/src/server.rs:255]: `.expect("There")` is not saying much @Laursen79

File: client/src/encryption.rs [hunk 25]
```diff
@@ -40,39 +52,85 @@ pub async fn decrypt<R: Rng + CryptoRng>(
...
+}
+
+pub fn unpad_message(message: &[u8]) -> Result<Vec<u8>, PaddingError> {
+    for i in 0..message.len() {
+        if message[i] == 0x80 {
+            return Ok(message[0..i].to_vec());
+        }
+    }
+    Err(PaddingError::UnpadError)
+}
```
Comment (andreasalstrup) [inline client/src/encryption.rs:106]: Hvorfor `0x80`?

Hvad hvis `0x80` er midt i en message?

File: client/src/errors.rs [hunk 26]
```diff
@@ -2,43 +2,105 @@ use std::error::Error;
 use std::fmt;
 use std::fmt::Display;
 
+use common::protocol_address::ParseProtocolAddressError;
+use libsignal_core::DeviceId;
+use libsignal_protocol::SignalProtocolError;
+
+type Result<T> = std::result::Result<T, SignalClientError>;
```
Comment (andreasalstrup) [inline client/src/errors.rs:9]: unused

File: client/src/errors.rs [hunk 27]
```diff
@@ -2,43 +2,105 @@ use std::error::Error;
...
+#[derive(Debug)]
+pub enum ServerRequestError {
+    /// The status code was not 200 - OK
+    StatusCodeError(u16, String),
+    BodyDecodeError(String),
+    TransmissionError(String),
 }
 
+impl Display for ServerRequestError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
```
Comment (andreasalstrup) [inline client/src/errors.rs:33]: Kan `std::fmt::Formatter` og `std::fmt::Result` importeres?

File: client/src/server.rs
Comment (andreasalstrup) [inline client/src/server.rs:1]: Vi kalder filen `server.rs` er det meningen? Burde nok hedde `backend.rs` ifølge navngivningen i filen.

Kunne også hedde noget andet (etc. `interface`) , synes `backend` er lidt misvisende. Men det er kun personal opinion

File: client/src/server.rs [hunk 28]
```diff
@@ -174,74 +233,92 @@ impl Server for ServerAPI {
...
+            .map_err(SendMessageError::WebSocketError)?;
 
-    async fn update_client(&self, client: &Contact) -> Result<(), Self::Error> {
-        let payload = json!({
-            "uuid": client.service_id.service_id_string()
-        });
-        self.make_request(ReqType::Put(payload), CLIENT_URI.to_string())
-            .await;
-        todo!("Handle the response")
+        // TODO handle send went wrong.
```
Comment (andreasalstrup) [inline client/src/server.rs:287]: Skal det fixes i dette PR?

File: client/src/server.rs [hunk 29]
```diff
@@ -174,74 +233,92 @@ impl Server for ServerAPI {
...
+        &mut self,
+        messages: SignalMessages,
+        recipient: &ServiceId,
+    ) -> Result<(), SignalClientError> {
+        let payload = to_vec(&messages).unwrap();
+        let uri = format!("{}/{}", MSG_URI, recipient.service_id_string());
+        println!("Sending message to: {}", uri);
+
+        let id = self.socket_manager.next_id();
+        let response = self
```
Comment (andreasalstrup) [inline client/src/server.rs:281]: unused

**Review by andreasalstrup — CHANGES_REQUESTED** [review]
Looks good, just some requested changes

File: client/src/client.rs (hunk 8 above)
Comment (Laursen79) [reply client/src/client.rs:246]: Maybe or should?

File: client/src/client.rs [hunk 30]
```diff
@@ -197,11 +204,296 @@ impl<T: StorageType> Client<T> {
...
+                .map_err(ReceiveMessageError::ProtobufDecodeContentError)?
+                .data_message
+                .ok_or_else(|| ReceiveMessageError::InvalidMessageContent)?
+                .body()
+                .to_owned(),
+        )
+    }
+}
+
+fn decode_ciphertext(
```
[the remark this answers is not in the document]
Comment (Laursen79) [reply client/src/client.rs:363]: mayhaps

File: client/src/client.rs [hunk 31]
```diff
@@ -197,11 +204,296 @@ impl<T: StorageType> Client<T> {
...
+    use common::signalservice::{Content, DataMessage};
+    use libsignal_core::{Aci, ProtocolAddress};
+    use libsignal_protocol::IdentityKeyPair;
+    use prost::Message;
+    use rand::rngs::OsRng;
+    use tokio::sync::Mutex;
+    use uuid::uuid;
+
+    /*
+    fn get_alice(state: Arc<Mutex<MockBackendState>>) -> Client<InMemory, MockBackend> {
```
[the remark this answers is not in the document]
Comment (Laursen79) [reply client/src/client.rs:409]: Does not work after keymanager refactor

File: client/src/encryption.rs (hunk 25 above)
Comment (Laursen79) [reply client/src/encryption.rs:106]: Det gør signal

File: client/src/server.rs (hunk 29 above)
Comment (Laursen79) [reply client/src/server.rs:279]: There is an issue on that #168

File: client/src/encryption.rs (hunk 6 above)
Comment (Laursen79) [reply client/src/encryption.rs]: Fixed?

File: client/src/client.rs [hunk 32]
```diff
@@ -1,53 +1,54 @@
...
-    }
+    pub storage: Storage<ST>,
 }
 
 const PROFILE_KEY_LENGTH: usize = 32;
 const MASTER_KEY_LENGTH: usize = 32;
 const PASSWORD_LENGTH: usize = 16;
 
-impl<T: StorageType> Client<T> {
+impl<S: StorageType, B: SignalServerAPI> Client<S, B> {
```
Comment (Mikkel-Boje) [inline client/src/client.rs:44]: Please have the same names for the generics

**Review by Mikkel-Boje — DISMISSED** [review]

**Review by andreasalstrup — DISMISSED** [review]

File: client/src/client.rs [hunk 33]
```diff
@@ -159,25 +146,29 @@ impl<T: StorageType> Client<T> {
...
 
-        server_api.publish_bundle(key_bundle).await?;
+        server_api.publish_pre_key_bundle(key_bundle).await?;
 
         server_api
-            .fetch_bundle(aci.service_id_string(), "1".to_string())
+            .connect(
+                &aci.service_id_string(),
+                &password,
+                "wss://127.0.0.1:4444/v1/websocket",
```
Comment (MJHC) [inline client/src/client.rs:169]: fix this

**Review by MJHC — CHANGES_REQUESTED** [review]

**Review by Darkros1245 — APPROVED** [review]

**Review by Darkros1245 — APPROVED** [review]

**Review by Mikkel-Boje — APPROVED** [review]

**Pull request merged**